File: python/qlogicae_cor/v1/error_manager.py
from qlogicae_cor.v1.error_manager_configurations import (
    ErrorManagerConfigurations,
)


class ErrorManager:
    __slots__ = (
        "_configurations"
    )

    def __init__(self) -> None:
        self._configurations: ErrorManagerConfigurations = ErrorManagerConfigurations()

    @property
    def configurations(self) -> ErrorManagerConfigurations:
        return self._configurations

    @configurations.setter
    def configurations(self, value: ErrorManagerConfigurations) -> None:
        self._configurations = value

    def setup(
        self,
        new_configurations: ErrorManagerConfigurations,
    ) -> bool:
        if self._configurations.is_disabled_for_handling(
            new_configurations is None,
        ):
            return False

        self._configurations = new_configurations

        return True

    def reset(
        self,
    ) -> bool:
        if self._configurations.is_disabled_for_handling():
            return False

        self._configurations = ErrorManagerConfigurations()

        return True

    def transform_to_error_log(
        self,
        value1: str | Exception,
        value2: str | None = None,
    ) -> str:
        if isinstance(value1, Exception):
            return (
                f"{type(value1).__name__}"
                f"{self._configurations.title_message_separator}"
                f"{value1}"
            )

        if value2 is not None:
            return f"{value1}{self._configurations.title_message_separator}{value2}"

        return (
            f"{self._configurations.title}"
            f"{self._configurations.title_message_separator}"
            f"{value1}"
        )

    def handle_error_outputs(
        self,
        error: str | Exception,
        message: str | None = None,
    ) -> bool:
        if isinstance(error, Exception):
            self.handle_error_output_conditions(
                self.transform_to_error_log(error),
            )

            return True

        if message is not None:
            self.handle_error_output_conditions(
                self.transform_to_error_log(
                    error,
                    message,
                ),
            )

            return True

        self.handle_error_output_conditions(
            self.transform_to_error_log(error),
        )

        return True

    def handle_error_output_conditions(
        self,
        error_log: str,
    ) -> bool:
        if not self._configurations.is_enabled_for_runtime_execution_handling():
            return False

        if self._configurations.is_enabled_for_console_output():
            print(error_log)

        if self._configurations.is_enabled_for_runtime_throw_output():
            raise RuntimeError(error_log)

        return True

File: cpp/qlogicae_cor1_utilities/qlogicae/cor1/sources/error_manager.cpp
#include "pch.hpp"

#if QLOGICAE_COR1__BASE__HPP_CPP__IS_COMPILATION_CONDITIONS_ENABLED_TEMPLATE( \
		ErrorManager \
	)

#include "../includes/error_manager.hpp"

namespace
	QLOGICAE_COR1__BASE__HPP_CPP__COR_NAMESPACE_NAME
{            
	ErrorManager
		::ErrorManager()
	{
		construct();	
	}
	
	ErrorManager
		::~ErrorManager()
	{
		destruct();		
	}

	bool
		ErrorManager
			::construct()
	{
		try
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__PRE_EXECUTION_GUARD_TEMPLATE
			(
				QLOGICAE_COR1__BASE__HPP_CPP__MUTEX_LAYER_1,
				QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
			);

			return
				true;
		}
		catch
		(
			QLOGICAE_COR1__BASE__HPP_CPP__TRY_CATCH_EXCEPTION_PARAMETER
		)
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__CATCH_CODE_TEMPLATE();
		}
	}

	bool
		ErrorManager
			::destruct()
	{
		try
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__PRE_EXECUTION_GUARD_TEMPLATE
			(
				QLOGICAE_COR1__BASE__HPP_CPP__MUTEX_LAYER_1,
				QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
			);

			return
				true;
		}
		catch
		(
			QLOGICAE_COR1__BASE__HPP_CPP__TRY_CATCH_EXCEPTION_PARAMETER
		)
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__CATCH_CODE_TEMPLATE();
		}
	}

	bool
		ErrorManager
			::setup(
				const ErrorManagerConfigurations&
					new_configurations
			)
	{
		try
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__PRE_EXECUTION_GUARD_TEMPLATE
			(
				QLOGICAE_COR1__BASE__HPP_CPP__MUTEX_LAYER_1,
				QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
			);

			configurations =
				new_configurations;

			return
				true;
		}
		catch
		(
			QLOGICAE_COR1__BASE__HPP_CPP__TRY_CATCH_EXCEPTION_PARAMETER
		)
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__CATCH_CODE_TEMPLATE();
		}
	}

	bool
		ErrorManager
			::reset()
	{
		try
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__PRE_EXECUTION_GUARD_TEMPLATE
			(
				QLOGICAE_COR1__BASE__HPP_CPP__MUTEX_LAYER_1,
				QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
			);

			configurations =
				{};

			return
				true;
		}
		catch
		(
			QLOGICAE_COR1__BASE__HPP_CPP__TRY_CATCH_EXCEPTION_PARAMETER
		)
		{
			QLOGICAE_COR1__IMPLICIT__HPP_CPP__CATCH_CODE_TEMPLATE();
		}
	}   

	std::string
		ErrorManager
			::transform_to_error_log(
				const std::string&
					title,
				const std::string&
					message
			)
	{
		QLOGICAE_COR1__IMPLICIT__HPP_CPP__EDGE_CASE_GUARD_TEMPLATE
		(
			QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
		);

		return
			title +
			configurations
				.title_message_separator +
			message;
	}


	std::string
		ErrorManager
			::transform_to_error_log(
				const std::string&
					message
			)
	{
		QLOGICAE_COR1__IMPLICIT__HPP_CPP__EDGE_CASE_GUARD_TEMPLATE
		(
			QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
		);

		return
			configurations
				.default_title +
			configurations
				.title_message_separator +
			message;
	}

	std::string
		ErrorManager
			::transform_to_error_log(
				const std::exception&
					exception
				)
	{
		QLOGICAE_COR1__IMPLICIT__HPP_CPP__EDGE_CASE_GUARD_TEMPLATE
		(
			QLOGICAE_COR1__BASE__HPP_CPP__EMPTY_EDGE_CASES
		);

		return
			configurations
				.default_title +
			configurations
				.title_message_separator +
			exception
				.what();
	}
}

#endif

